network/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Cumulus-specific network implementation.
+//!
+//! Contains message send between collators and logic to process them.
+
+use substrate_client::error::{Error as ClientError};
+use sr_primitives::traits::{Block as BlockT};
+use substrate_consensus_common::block_validation::{Validation, BlockAnnounceValidator};
+
+use polkadot_primitives::{
+	Hash as PHash, parachain::{CandidateReceipt, ValidatorIndex, ValidatorSignature, ValidatorId},
+};
+use polkadot_statement_table::Statement;
+use polkadot_validation::check_statement;
+
+use codec::{Decode, Encode};
+
+use std::marker::PhantomData;
+
+/// Justification that a parachain block is the parachain block candidate of one of the relay chain
+/// validator.
+#[derive(Encode, Decode)]
+pub struct BlockCandidateJustification {
```

Comment:
Why not use the polkadot candidate message directly?

Do you want to save some spaces? If yes, could you please provide a `From` implementation.

Reply:
or maybe you mean using a field with Statement instead of CandidateReceipt. in this case ok yes I'll do it, It is better indeed.

Reply:
I just thought that the whole justification should just be the statement, as the statement should contain all of these information?

Reply:
`polkadot-network::GossipStatement` contains all informations. we can make this being the justification.

Reply:
I mean we can reuse this entirely and we just need to check header match as well.

https://github.com/paritytech/polkadot/blob/master/network/src/gossip.rs#L426-L455

But I feel like polkadot could change a bit and require more than this to check the message, and this would be unfortunate.

I can make this polkadot stuff more public and reuse it here. I thought it is better to not rely on the whole stuff as polkadot could change MessageValidationData and if it happens cumulus can't have some of them then we are stuck back to this current state.

Reply:
But polkadot can also not change these structs as will, as they are specified as well.